changes/2024-TBD_cache-across-hierarchy-keyrings/change.md

```diff
@@ -0,0 +1,438 @@
+| .NET     | 1.4.0              | n/a             | [CreateAwsKmsHierarchicalKeyringInput.cs](https://github.com/aws/aws-cryptographic-material-providers-library/blob/main/AwsCryptographicMaterialProviders/runtimes/net/Generated/AwsCryptographicMaterialProviders/CreateAwsKmsHierarchicalKeyringInput.cs)                                         |
+
+### Conventions used in this document
+
+The key words
+"MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
+"SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL"
+in this document are to be interpreted as described in
+[RFC 2119](https://tools.ietf.org/html/rfc2119).
+
+## Definitions
+
+An "MPL Consumer" is a library, service, or other application
+that uses the AWS Cryptographic Material Providers Library (MPL)
+to manage cryptographic materials.
+An "MPL Consumer" MAY be using an AWS Crypto Tools product,
+such as the AWS Encryption SDK or AWS Database Encryption SDK.
+
+By "KMS Relationship", we mean any or all of the following:
+
+- KMS Configuration
+- Credentials used when creating the KMS Client, and thus used when calling KMS
+- Other properties of the KMS Client, such as the region, or request headers
+
+## Background
+
+A [Cryptographic Materials Cache](https://github.com/awslabs/aws-encryption-sdk-specification/blob/master/framework/cryptographic-materials-cache.md) can be used by a Hierarchy Keyring, a Caching CMM, and the DB-ESDK.
+It can store four types of [materials](https://github.com/aws/aws-cryptographic-material-providers-library/blob/c07a51fc29ff70411f7573bca96d2a091db8c1ed/AwsCryptographicMaterialProviders/dafny/AwsCryptographicMaterialProviders/Model/cryptographic-materials-cache.smithy#L89) for three different use-cases:
+
+- Hierarchy Keyring: keyStore#BranchKeyMaterials
+- CachingCMM: EncryptionMaterials and DecryptionMaterials
+- DB-ESDK: keyStore#BeaconKeyMaterials
+
+These materials have certain cache identifiers for accessing
+them until a TTL. As of [today](#today), the cache identifiers
+are set up such that:
+
+- Only one Hierarchy Keyring can use one cache,
+  i.e., currently, we do NOT support a shared cache
+  between multiple Hierarchy Keyrings.
+- Multiple CachingCMMs can use a single shared cache
+  (this is for native languages, there is currently no
+  CachingCMM implemented in dafny)
+- We do NOT support a shared cache between multiple
+  DB-ESDK clients.
+
+We will focus on the Hierarchy Keyring unless mentioned
+otherwise. Our aim is to support a shared cache between
+multiple Hierarchy Keyrings
+as highlighted in the next sections.
+
+## Motivation
+
+The Hierarchy Keyring,
+and it's component the (Branch) Keystore,
+allow MPL Consumers to reduce their KMS Call volume
+by persisting KMS protected cryptographic materials into
+an available medium
+(currently, only a DynamoDB table is available as persistence medium).
+
+We call these cryptographic materials Branch Keys.
+
+However, an instance of the Hierarchy Keyring
```

Comment:
does this mean, multiple keyrings may have the same KMS relationship?

Reply:
Yes, you can create multiple keyrings with the same KMS relationships.

Reply:
so then wouldn't multiple keyrings (sharing the same keystore) create cache IDs with collision?